### ISSUE TITLE
OCP 4.12: some tests fail with PodSecurity restrictions error

### ISSUE THREAD

**Issue by nyechiel** [submariner-io/submariner-operator#2317]
- `subctl`-based deployment on top of OCP 4.12 (k8s 1.25) clusters using `0.14.0-rc1`
- Installation seems to be working fine; IPsec connections are up
- `subctl diagnose` shows pod security warnings as described here: https://github.com/submariner-io/subctl/issues/348
- Some E2E tests are failing with what seems to be wrong PodSecurity configuration:

```
• Failure [1.023 seconds]
[dataplane-globalnet] Basic TCP connectivity tests across overlapping clusters without discovery
github.com/submariner-io/submariner@v0.14.0-rc1/test/e2e/dataplane/tcp_gn_pod_connectivity.go:28
  when a pod connects via TCP to the globalIP of a remote service in reverse direction
  github.com/submariner-io/submariner@v0.14.0-rc1/test/e2e/dataplane/tcp_gn_pod_connectivity.go:154
    when the pod is not on a gateway and the remote service is not on a gateway
    github.com/submariner-io/submariner@v0.14.0-rc1/test/e2e/dataplane/tcp_gn_pod_connectivity.go:163
      should have sent the expected data from the pod to the other pod [It]
      github.com/submariner-io/submariner@v0.14.0-rc1/test/e2e/dataplane/tcp_gn_pod_connectivity.go:37

      Unexpected error:
          <*errors.StatusError | 0xc0009c7040>: {
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "pods \"tcp-check-listener4rjvd\" is forbidden: violates PodSecurity \"restricted:v1.24\": allowPrivilegeEscalation != false (container \"tcp-check-listener\" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container \"tcp-check-listener\" must set securityContext.capabilities.drop=[\"ALL\"]), runAsNonRoot != true (pod or container \"tcp-check-listener\" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container \"tcp-check-listener\" must set securityContext.seccompProfile.type to \"RuntimeDefault\" or \"Localhost\")",
                  Reason: "Forbidden",
                  Details: {
                      Name: "tcp-check-listener4rjvd",
                      Group: "",
                      Kind: "pods",
                      UID: "",
                      Causes: nil,
                      RetryAfterSeconds: 0,
                  },
                  Code: 403,
              },
          }
          pods "tcp-check-listener4rjvd" is forbidden: violates PodSecurity "restricted:v1.24": allowPrivilegeEscalation != false (container "tcp-check-listener" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "tcp-check-listener" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "tcp-check-listener" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "tcp-check-listener" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
      occurred

      github.com/submariner-io/shipyard@v0.14.0-rc1/test/e2e/framework/network_pods.go:283
```

See attached files.

**Comment by nyechiel** [comment]
[full_e2e_output.txt](https://github.com/submariner-io/submariner-operator/files/9904099/full_e2e_output.txt)

[nyechiel-ovnk-1.zip](https://github.com/submariner-io/submariner-operator/files/9904112/nyechiel-ovnk-1.zip)

[nyechiel-ovnk-2.zip](https://github.com/submariner-io/submariner-operator/files/9904114/nyechiel-ovnk-2.zip)

**Comment by nyechiel** [comment]
This is likely related to https://github.com/submariner-io/shipyard/issues/1006

**Comment by tpantelis** [comment]
With https://github.com/submariner-io/shipyard/pull/1009, I set `allowPrivilegeEscalation`  and `Capabilities.Drop` according to the error message and to match what was previously done in https://github.com/submariner-io/shipyard/pull/995/files.

Also, according to the error, we "_must set securityContext.runAsNonRoot=true_". But according to [this](https://master.sdk.operatorframework.io/docs/best-practices/pod-security-standards/):

```
# WARNING: Ensure that the image used defines an UserID in the Dockerfile
# otherwise the Pod will not run and will fail with `container has runAsNonRoot and image has non-numeric user`.
# If you want your workloads admitted in namespaces enforced with the restricted mode in OpenShift/OKD vendors
# then, you MUST ensure that the Dockerfile defines a User ID OR you MUST leave the `RunAsNonRoot` and
# RunAsUser fields empty.
 runAsNonRoot: true
```

So not sure we can use this setting.

Also the error says to "_set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost"_" but the article  mentioned above states:

```
# Please ensure that you can use SeccompProfile and do not use
# if your project must work on old Kubernetes
# versions < 1.19 or on vendors versions which
# do NOT support this field by default (i.e. Openshift < 4.11 ) 
```

which would be problematic for us.

So hopefully setting `allowPrivilegeEscalation`  and `Capabilities.Drop` will suffice.

**Comment by skitt** [comment]
> With [submariner-io/shipyard#1009](https://github.com/submariner-io/shipyard/pull/1009), I set `allowPrivilegeEscalation` and `Capabilities.Drop` according to the error message and to match what was previously done in https://github.com/submariner-io/shipyard/pull/995/files.
> 
> Also, according to the error, we "_must set securityContext.runAsNonRoot=true_". But according to [this](https://master.sdk.operatorframework.io/docs/best-practices/pod-security-standards/):
> 
> ```
> # WARNING: Ensure that the image used defines an UserID in the Dockerfile
> # otherwise the Pod will not run and will fail with `container has runAsNonRoot and image has non-numeric user`.
> # If you want your workloads admitted in namespaces enforced with the restricted mode in OpenShift/OKD vendors
> # then, you MUST ensure that the Dockerfile defines a User ID OR you MUST leave the `RunAsNonRoot` and
> # RunAsUser fields empty.
>  runAsNonRoot: true
> ```
> 
> So not sure we can use this setting.

In my tests, the pods run fine with this setting on OCP 4.12.

> Also the error says to "_set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost"_" but the article mentioned above states:
> 
> ```
> # Please ensure that you can use SeccompProfile and do not use
> # if your project must work on old Kubernetes
> # versions < 1.19 or on vendors versions which
> # do NOT support this field by default (i.e. Openshift < 4.11 ) 
> ```
> 
> which would be problematic for us.

Yes, we’d have to adjust the specs depending on the version of K8s on the cluster.

> So hopefully setting `allowPrivilegeEscalation` and `Capabilities.Drop` will suffice.

Unfortunately it isn’t.

**Comment by tpantelis** [comment]
> In my tests, the pods run fine with this setting on OCP 4.12.
> 

hmm... but [not in kind](https://github.com/submariner-io/shipyard/actions/runs/3377773219/jobs/5607117483): `container has runAsNonRoot and image will run as root`. I assume that's b/c the user ID defaults to 0. So we need to set it to something.

**Comment by sridhargaddam** [comment]
The PR (https://github.com/submariner-io/shipyard/pull/1009/files) seems to be merged. So, can we close it or is there any additional work pending for this issue?

**Comment by tpantelis** [comment]
> The PR (https://github.com/submariner-io/shipyard/pull/1009/files) seems to be merged. So, can we close it or is there any additional work pending for this issue?

Need to backport still...

**Comment by nyechiel** [comment]
Once we confirm the fix, let's also backport to `release-0.13`. I added this issue into https://github.com/submariner-io/releases/issues/525
